### PR TITLE
wip: initial poc to add WASM backed UDFs

### PR DESCRIPTION
**Pull request by nathanielc**
Closes #

Describe your proposed changes here.

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

**Comment by crepererum** [comment]
I think we have two ways to implement UDFs in WASM:

1. **scalar function calls:** Call the WASM function for every scalar/tuple/row. That's what you do here. Easy to implement (esp. for the user), but will probably not scale very well because you have a method call overhead per row.
2. **vector function calls:** Call the WASM function with arrays for every column. That's slightly more difficult to implement for the user (requires an additional loop), but the overhead will be way smaller. Also the WASM runtime could perform vectorization of loops.
3. **arrow function calls:** Allow the WASM function to retrieve/send arrow data structures directly. This has max. complexity for the user, but we could provide frameworks/tools to simplify this (maybe even adapters that call variant 1 and 2). See [WASM Arrow](https://github.com/domoritz/arrow-wasm/) for prior art.

File: iox_query/src/frontend/wasm.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+use std::sync::Arc;
+
+use crate::exec::context::IOxSessionContext;
+use datafusion::{error::Result, physical_plan::ExecutionPlan};
+
+/// This struct can create plans for running SQL queries against databases
```
Comment (alamb) [inline iox_query/src/frontend/wasm.rs:6]: You probably already know this but I don't think you need this specific frontend -- you can probably just use the normal SQL frontend

File: wasm_demo/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#[no_mangle]
+pub extern "C" fn wasm_mul(a: f64, b: f64) -> f64 {
```
Comment (alamb) [inline wasm_demo/src/lib.rs:2]: 👍  nice

**Review by alamb — COMMENTED** [review]
This is very cool -- I won't say I followed all the wasm trampoline code but the basic idea looks awesome

**Pull request closed without merging**